feat(jest-expo): add experimental RSC testing library (#29404)
expo/expo: docs/pages/guides/testing-rsc.mdx

---
title: Testing React Server Components
sidebar_title: Testing RSC

> description: Learn about writing unit tests for React and React Native Server Components in Expo.

Reviewer: The description mentions "React" here. Is it different from using Unit testing in general (the instructions we mention here: https://docs.expo.dev/develop/unit-testing/)?

Author: It's like unit testing if the tests ran on an iOS device instead of in the terminal.

platforms: ['android', 'ios', 'web']
---

import { Terminal } from '~/ui/components/Snippet';

> > This guide refers to the experimental feature React Server Components which is still in development.

Reviewer: ```suggestion
> **warning** This guide refers to the experimental feature React Server Components which is still in development.
```

Let's change this callout to warning.

Expo has support for testing React Server Components in your project.

> ## Jest Testing

Reviewer: ```suggestion
## Jest testing
```

> Unlike client-based Jest testing, React Server Components run in a Node.js environment, meaning you can more closely emulate the server-side rendering environment with Jest on its own.

Reviewer: ```suggestion
React Server Components run in a Node.js environment, unlike a client-based Jest testing environment. This means you can emulate the server-side rendering environment closely with Jest.
```

Let's rephrase the above paragraph for following reasons:
- Starting with a positive note always helps when reading
- Let's split sentences and rephrase the second sentence for brevity

### Setup

> While standard server rendering is web-only, Expo's universal React Server Components bundle custom server renderers for each platform. This means platform-extensions like `*.ios.js` and `*.native.ts` will be resolved when writing server components for an iOS app.

Reviewer: ```suggestion
While standard server rendering is web-only, Expo's universal RSC bundles custom server renderers for each platform. For example, when writing server components for an iOS app, the platform-specific extensions such as **\*.ios.js** and **\*.native.ts** will be resolved.
```

- Fix grammatical error
- We can use the short form for React Server Components since that is already defined in the introduction for brevity.
- The second sentence targets only one native platform, so it portrays an example here. I've updated the above suggestion. However, if we want to portray both Android and iOS, we can rephrase it as:

```md
When writing server components for an Android or iOS app, the platform-specific extensions such as **\*.android.js**, **\*.ios.js**, and **\*.native.ts** will be resolved.
```

`jest-expo` provides a couple different presets for testing server components:

> - `jest-expo/rsc/web`: An Android-only runner for React Server Components. Uses `*.web.js` and `*.js` files.
> - `jest-expo/rsc/ios`: An iOS-only runner for React Server Components. Uses `*.ios.js`, `*.native.js`, and `*.js` files.
> - `jest-expo/rsc/android`: An Android-only runner for React Server Components. Uses `*.android.js`, `*.native.js`, and `*.js` files.
> - `jest-expo/rsc`: A multi-runner which combines the above runners.

Reviewer: ```suggestion
| Runner                  | Description                                                                                                        |
| ----------------------- | ------------------------------------------------------------------------------------------------------------------ |
| `jest-expo/rsc/android` | An Android-only runner for RSC. Uses **\*.android.js**, **\*.native.js**, and **\*.js** files. |
| `jest-expo/rsc/ios`     | An iOS-only runner for RSC. Uses **\*.ios.js**, **\*.native.js**, and **\*.js** files.         |
| `jest-expo/rsc/web`     | A web-only runner for RSC. Uses **\*.web.js** and **\*.js** files.                            |
| `jest-expo/rsc`         | A multi-runner that combines the above runners.                                                                   |
```

Let's convert the above list items into a table to improve readability with the following changes
- There's a typo in the first item. It's for web.
- Let's use bold filenames as per our writing style guide.
- Use correct platform-order (Android, iOS, web) as per our writing style guide
- We can use the short form for React Server Components since that is already defined in the introduction for brevity.

> We recommend creating a `jest-rsc.config.js` file in your project root to configure Jest for React Server Components:

Reviewer: ```suggestion
To configure Jest for RSC, create a **jest-rsc.config.js** file in your project's root:
```

- Let's rephrase this and use bold filenames.
- We can use short form for React Server Components since that is already defined in introduction, for brevity.

```js jest-rsc.config.js
module.exports = require('jest-expo/rsc/jest-preset');
```

> Then you can add a script like `test:rsc` to your `package.json`:

Reviewer: ```suggestion
Then, you can add a script such as `test:rsc` to your **package.json**:
```

```json package.json
{
  "scripts": {
    "test:rsc": "jest --config jest-rsc.config.js"
  }
}
```

### Writing tests

> Tests should be written in a `__rsc_tests__` directory to prevent Jest from running your client tests on the server.

Reviewer: ```suggestion
Tests should be written in a **\_\_rsc_tests\_\_** directory to prevent Jest from running your client tests on the server.
```

```js __rsc_tests__/my-component.test.ts
/// <reference types="jest-expo/rsc/expect" />

import { LinearGradient } from 'expo-linear-gradient';

it(`renders to RSC`, async () => {
  const jsx = (
    <LinearGradient
      colors={['cyan', '#ff00ff', 'rgba(0,0,0,0)', 'rgba(0,255,255,0.5)']}
      testID="gradient"
    />
  );

  await expect(jsx).toMatchFlight(`1:I["src/LinearGradient.tsx",[],"LinearGradient"]
0:["$","$L1",null,{"colors":["cyan","#ff00ff","rgba(0,0,0,0)","rgba(0,255,255,0.5)"],"testID":"gradient"},null]`);
});
```

> Any code you import in your tests will be run in the server environment, so you can import server-only modules like `react-server` and `server-only`. This is useful for determining if a library is compatible with React Server Components.

Reviewer: ```suggestion
Any code you import in your test files will run in the server environment. You can import server-only modules like `react-server` and `server-only`. This is useful for determining if a library is compatible with RSC.
```

We can use short form for React Server Components since that is already defined in introduction, for brevity.

### Custom expect matchers

Jest Expo for RSC adds a couple of custom matchers to Jest's expect:

- `toMatchFlight`: Render a JSX element using a pseudo-implementation of the render in Expo CLI and compare to a flight string.
- `toMatchFlightSnapshot`: Same as `toMatchFlight` but saves the flight string to a snapshot file.

> Behind the scenes, these methods are performing part of the framework operation required to render React Server Components. The component is streamed to a string and compared all at once. You can alternatively stream it manually to observe the progress.

Reviewer: ```suggestion
Behind the scenes, these methods handle a part of the framework operation needed to render RSC. The component is streamed to a string and compared all at once. You can alternatively stream it manually to observe the progress.
```

> If a component fails to render, the matcher will throw an error to fail the test. In reality, an `E:` line will be present in your flight string, this will be sent to the client where the error will end up being thrown.

Reviewer: ```suggestion
If a component fails to render, the matcher will throw an error to fail the test. An `E:` line will be present in your flight string in real time. This will be sent to the client, where the error will be thrown.
```

### Running tests

You can run your tests with the `test:rsc` script:

> ```sh
> yarn test:rsc --watch
> ```

Reviewer: ```suggestion
<Terminal cmd={['$ yarn test:rsc --watch']} />
```

Let's replace this command block with `Terminal` component.

If you're using the multi-runner, you can select a specific project using the `--selectProjects` flag. The following example only runs the web platform:

> ```sh
> yarn test:rsc --watch --selectProjects rsc/web
> ```

Reviewer: ```suggestion
<Terminal cmd={['$ yarn test:rsc --watch --selectProjects rsc/web']} />
```

### Environments

> In a React Server Components bundling environment, you can import files like

Reviewer: ```suggestion
In an RSC bundling environment, you can import files like
```

There seems to be some text missing here.

> If a component fails to render, the matcher will throw an error to fail the test. An `E:` line will be present in your flight string in real time. This will be sent to the client, where the error will be thrown.

Author: ```suggestion
If a component fails to render, the matcher will throw an error to fail the test. In practice, the server renderer will generate an `E:` line, which will sent to the client to be thrown locally for the user.
```

## Tips

Use the `server-only` and `client-only` modules to assert that a module should not be imported on the client or server:

```js my-module.js
import 'server-only';
```

> React Server Components support package exports by default, you can use the `react-server` condition to change what file is imported from a module:

Reviewer: ```suggestion
RSC supports package exports by default. You can use the `react-server` condition to change what file is imported from a module:
```

```json package.json
{
  "exports": {
    ".": {
      "react-server": "./index.react-server.js",
      "default": "./index.js"
    }
  }
}
```

> When a process is bundling for React Server Components, all modules are bundled in React Server-mode and you opt-out with the `"use client"` directive. When `"use client"` is found, the module becomes an async reference to the client module. `"use server"` is not the opposite of `"use client"`, it is instead used to define React Server Actions.

Reviewer: ```suggestion
When a process is bundling for RSC, all modules are bundled in React Server mode and you can opt-out with the `"use client"` directive. When `"use client"` is found, the module becomes an async reference to the client module. `"use server"` is not the opposite of `"use client"`. It is instead used to define React Server Actions.
```

platforms: ['android', 'ios', 'web']
---

import { Terminal } from '~/ui/components/Snippet';

> This guide refers to the experimental feature React Server Components which is still in development.

React Server Components (RSC) is a new feature in React that allows you to build components that render on the server and can be hydrated on the client. This guide provides details on how to write unit tests for RSC in your project.

## Jest testing

React Server Components run in a Node.js environment, unlike a client-based Jest testing environment. This means you can emulate the server-side rendering environment closely with Jest.

### Setup

> While standard server rendering is web-only, Expo's universal RSC bundles custom server renderers for each platform. For example, when writing server components for an iOS app, the platform-specific extensions such as **\*.ios.js** and **\*.native.ts** will be resolved.

Reviewer 2: The "For example" is overly broad. Just the iOS part is the example. The "the" sounds extraneous to me and reads better without it.

```suggestion
While standard server rendering is web-only, Expo's universal RSC bundles custom server renderers for each platform. This means platform-specific file extensions are supported. For example, when writing server components for an iOS app, platform-specific extensions such as **\*.ios.js** and **\*.native.ts** will be resolved.
```

});
```

Any code you import in your test files will run in the server environment. You can import server-only modules like `react-server` and `server-only`. This is useful for determining if a library is compatible with RSC.

### Custom expect matchers

`jest-expo` for RSC adds a couple of custom matchers to Jest's `expect`:

- `toMatchFlight`: Render a JSX element using a pseudo-implementation of the render in Expo CLI and compare to a flight string.
- `toMatchFlightSnapshot`: Same as `toMatchFlight` but saves the flight string to a snapshot file.

> Behind the scenes, these methods handle a part of the framework operation needed to render RSC. The component is streamed to a string and compared all at once. You can alternatively stream it manually to observe the progress.

Reviewer 2: I think the key word here is "buffer" rather than "stream". The point is that something collects the whole output, converting it from a stream to a complete datum.

```suggestion
Behind the scenes, these methods handle a part of the framework operation needed to render RSC. The component's render stream is buffered to a string and compared all at once. You can alternatively stream it manually to observe the rendering progress.
```

> If a component fails to render, the matcher will throw an error to fail the test. An `E:` line will be present in your flight string in real time. This will be sent to the client, where the error will be thrown.

Reviewer 2: I don't think "in real time" makes sense here

}
```

> When a process is bundling for RSC, all modules are bundled in React Server mode and you can opt-out with the `"use client"` directive. When `"use client"` is found, the module becomes an async reference to the client module. `"use server"` is not the opposite of `"use client"`. It is instead used to define React Server Actions.

Reviewer 2: "a process" is imprecise. I'd either replace it with a more precise noun or leave it out altogether and say, "When bundling for RSC".

De-hyphenate "opt out", it's not an adjective here.

```suggestion
When bundling for RSC, all modules are bundled in React Server mode and you can opt out with the `"use client"` directive. When `"use client"` is found, the module becomes an async reference to the client module.

`"use server"` is not the opposite of `"use client"`. It is instead used to define React Server Actions.
```
